pandas/core/case_when.py

```diff
@@ -0,0 +1,216 @@
+        Scalar,
+        Series,
+    )
+
+
+def case_when(
+    caselist: list[tuple[ArrayLike, ArrayLike | Scalar]],
+    default: ArrayLike | Scalar = lib.no_default,
+) -> Series:
+    """
+    Replace values where the conditions are True.
+
+    Parameters
+    ----------
+    caselist : List of tuples of conditions and expected replacements.
+        Takes the form:  ``(condition0, replacement0)``,
+        ``(condition1, replacement1)``, ... .
+        ``condition`` should be a 1-D boolean array.
+        When multiple boolean conditions are satisfied,
+        the first replacement is used.
+        If ``condition`` is a Series, and the equivalent ``replacement``
+        is a Series, they must have the same index.
+        If there are multiple replacement options,
+        and they are Series, they must have the same index.
+
+    default : scalar, array-like, default None
+        If provided, it is the replacement value to use
+        if all conditions evaluate to False.
+        If not specified, entries will be filled with the
+        corresponding NULL value.
+
+        .. versionadded:: 2.2.0
+
+    Returns
+    -------
+    Series
+
+    See Also
+    --------
+    Series.mask : Replace values where the condition is True.
+
+    Examples
+    --------
+    >>> df = pd.DataFrame({
+    ...     "a": [0,0,1,2],
+    ...     "b": [0,3,4,5],
+    ...     "c": [6,7,8,9]
+    ... })
+    >>> df
+       a  b  c
+    0  0  0  6
+    1  0  3  7
+    2  1  4  8
+    3  2  5  9
+
+    >>> caselist = [(df.a.gt(0), df.a), (df.b.gt(0), df.b)]  # condition, replacement
+    >>> pd.case_when(caselist=caselist, default=df.c)        # default is optional
+    0    6
+    1    3
+    2    1
+    3    2
+    Name: c, dtype: int64
+    """
+    from pandas import Series
+
+    validate_case_when(caselist=caselist)
+
+    conditions, replacements = zip(*caselist)
+    common_dtypes = [infer_dtype_from(replacement)[0] for replacement in replacements]
+
+    if default is not lib.no_default:
+        arg_dtype, _ = infer_dtype_from(default)
+        common_dtypes.append(arg_dtype)
+    else:
+        default = None
+    if len(set(common_dtypes)) > 1:
+        common_dtypes = find_common_type(common_dtypes)
+        updated_replacements = []
+        for condition, replacement in zip(conditions, replacements):
+            if is_scalar(replacement):
+                replacement = construct_1d_arraylike_from_scalar(
+                    value=replacement, length=len(condition), dtype=common_dtypes
+                )
+            elif isinstance(replacement, ABCSeries):
+                replacement = replacement.astype(common_dtypes)
+            else:
+                replacement = pd_array(replacement, dtype=common_dtypes)
+            updated_replacements.append(replacement)
+        replacements = updated_replacements
+        if (default is not None) and isinstance(default, ABCSeries):
+            default = default.astype(common_dtypes)
```

Comment:
cc @MarcoGorelli 

This might upcast, thoughts related to PDEP6?

Reply:
I don't think this is related to PDEP6, we are creating a new Series and not doing something like setitem.